### PR TITLE
Add specializing inline

### PR DESCRIPTION
**Pull request by odersky**
This PR just adds `<: T` return type syntax for inline methods and handles upcasting properly. The body of an `inline` method is upcasted to its return type, provided a return type is given with `: T`
(as opposed to `<: T`). 

We previously did not do this which meant that all inline methods had some degree of whiteboxity even though it was claimed otherwise.

This PR is based on #5382.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/Compiler.scala [hunk 1]
```diff
@@ -44,7 +44,7 @@ class Compiler {
   /** Phases dealing with TASTY tree pickling and unpickling */
   protected def picklerPhases: List[List[Phase]] =
     List(new Pickler) ::            // Generate TASTY info
-    List(new Staging) ::            // Inline calls, expand macros and turn quoted trees into explicit run-time data structures
+    List(new Staging) ::            // Turn quoted trees into explicit run-time data structures
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/Compiler.scala]: `Expand macros and turn ...`

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 2]
```diff
@@ -157,19 +157,31 @@ object desugar {
...
    *
    *      def f[T: B](x: Int = 1)(y: String = x + "m") = ...
    *  ==>
    *      def f[T](x: Int)(y: String)(implicit evidence$0: B[T]) = ...
    *      def f$default$1[T] = 1
    *      def f$default$2[T](x: Int) = x + "m"
+   *
+   *  3. Convert <: T to : T in specializing inline methods. E.g.
+   *
+   *      inline def f(x: Boolean) <: Any = if (x) 1 else ""
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Maybe we should encourage to write such signatures as union types. 

```scala
inline def f(x: Boolean) <: Int | String = if (x) 1 else ""
// or 
inline def f(x: Boolean) <: 1 | "" = if (x) 1 else ""
```

File: compiler/src/dotty/tools/dotc/ast/Trees.scala [hunk 3]
```diff
@@ -511,11 +511,18 @@ object Trees {
   }
 
   /** selector match { cases } */
-  case class Match[-T >: Untyped] private[ast] (selector: Tree[T], cases: List[CaseDef[T]])
+  case class Match[-T >: Untyped] private[ast] (selector: Tree[T], cases: List[CaseDef[T]])(val kind: MatchKind)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/ast/Trees.scala]: Do we really want `Match(s, cs)(Regular) == Match(s, cs)(Inline)`?

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: We can do that, but often it's more complicated than that.

File: compiler/src/dotty/tools/dotc/ast/Trees.scala (hunk 3 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/Trees.scala]: Tree `==` is identity, so this won't be a problem.

**Pull request merged**